### PR TITLE
Gate deletion/rename-aware "known command" checks across W123, arity, and dynamic-dispatch passes

### LINKED ISSUE #973: W123 doesn't fire for a call through a proc/class/alias/rename target deleted earlier in the file
## Summary

`Analyser::finalise_invocation_resolutions` (`rust/tcl-compiler/src/analyser/scope.rs`) settles each recorded invocation's resolved qualified name by testing candidate names against a shared `known()` predicate:

```rust
let known = |qualified: &str| {
    procs.contains_key(qualified)
        || classes.contains_key(qualified)
        || aliases.contains_key(qualified)
        || renames.contains_key(qualified)
        || (builtins.contains(qualified.trim_start_matches(':')) && !renamed_away.contains(qualified))
};
```

The registry-builtin clause explicitly excludes a name that was renamed/deleted away (`!renamed_away.contains(qualified)`, where `renamed_away` unions `renamed_commands` values with `deleted_commands` keys). The `procs`/`classes`/`aliases`/`renames` clauses have no equivalent gate — a name is treated as "known" purely because it was *ever* declared in the file, regardless of whether a later `rename NAME {}` (or an `interp alias` deletion) made it uncallable.

## Reproducer

```tcl
namespace eval ::a {
    proc helper {} { return 1 }
}
rename ::a::helper {}
namespace eval ::a {
    proc caller {} { helper }
}
```

`rename ::a::helper {}` makes `::a::helper` an invalid command name in real Tcl — any later call to it fails with `invalid command name "::a::helper"`. The `helper` call inside `caller` should therefore be unresolved (W123), matching what already happens for a *builtin* renamed away the same way. Instead, no W123 is emitted: `known("::a::helper")` returns `true` because `procs.contains_key("::a::helper")` is `true` — `all_procs` records the proc's declaration and is never pruned or gated when the command is later deleted.

Confirmed via the project's own analyser test harness — this reproduces identically for an ordinary bareword call as shown above, and is not specific to any one invocation kind, since a class, an alias target, and a rename target all share the same ungated `known()` arms.

## Why it matters

- W123 silently under-fires for a call that would actually error at runtime, which is the diagnostic's entire purpose.
- The same `resolved_qualified_name`/`resolution_candidates` fields feed go-to-definition, find-references, rename, call-hierarchy, minify, and completion, so each of those may also point at a dead command instead of correctly reporting nothing (or falling through to whatever *does* still resolve the name, if anything does).
- The registry-builtin arm shows the intended shape of the fix already exists for one of the four cases; the other three read like an oversight rather than a deliberate scoping decision — there's no comment explaining the asymmetry.

## Suggested fix direction

Gate the `procs`/`classes`/`aliases`/`renames` clauses in `known()` the same way the builtin clause already is. `Analyser::deleted_commands` already carries deletion offsets and is already consulted for order-gating in `unresolved.rs` (`qualified_name_deleted_before`, `registry_name_deleted_before`), and the same map is already read in `scope.rs` itself (used to build `renamed_away`). Two subtleties the fix needs to preserve, matching the existing order-gating convention elsewhere in the file:

- A later re-declaration of the same qualified name after the deletion (a fresh `proc`, a fresh `rename ... NAME`) must still count as known — this is a whole-file existence question, not a per-call-site one, matching how `alias_names`/`rename_target_names` are already built in `build_w123_known_names` (`unresolved.rs`).
- A deletion recorded from inside a proc/method/class body is conditional (it only executes if that body ever runs), so it should not disqualify the declaration the way a top-level deletion does — again matching the existing convention (`Analyser::offset_is_inside_definition_body`).

## Required tests

TP/FP/TN unit coverage (alongside the existing resolution tests in `scope.rs`/`commands.rs`, or in `analyser/diagnostics/tests.rs`) plus a native `lsp_e2e` diagnostics test, at minimum:

- a bareword call through a proc renamed away earlier in the file still draws W123 (the reproducer above);
- the same for a class, an `interp alias`, and a `rename` target;
- a proc re-declared under the same qualified name *after* the deletion still resolves (no W123);
- a deletion recorded inside a proc/class body that never runs does not disqualify the declaration (matching the existing conditional-deletion convention already used for the builtin case).

## Context

Found while auditing `finalise_invocation_resolutions` during the follow-up work on #968 (the `expr` math-function namespace-resolution fix). Confirmed generic and pre-existing by reproducing the identical non-detection with a plain ordinary command with no math function involved — it is unrelated to that fix and predates it.

### LINKED ISSUE #1006: W123 alias/rename-target checks use file-end-only deletion gating, not call-site/conditional-aware (follow-up to #973)
## Background

#973 fixed W123 ("unknown command") never firing for a proc/class/rename target/`interp alias` target that was renamed or deleted away with no later re-establishment. For procs and classes, the fix (`fact_live_for_call` in `rust/tcl-compiler/src/analyser/diagnostics/unresolved.rs`) is call-site- and conditional-body-aware: it checks whether the call is textually before an unconditional top-level deletion, and whether the deletion itself sits inside a proc/class/method body that might never execute — matching the sophistication `qualified_name_deleted_before` already gives registry builtins.

The alias (`alias_names`) and rename-target (`rename_target_names`) sets in `build_w123_known_names` still use the older, cruder `fact_live_at_file_end` check — file-end granularity only, no call-site or conditional-body awareness. This predates #973 (it's the mechanism #973's fix was modeled after) and was left untouched deliberately, since #973 was scoped to procs/classes. It carries two false positives that the proc/class path no longer has.

## The gaps (confirmed against tclsh 8.6.14)

**1. A call to an alias/rename-target textually *before* its later deletion still fires W123**, even though the call resolves fine at the time it runs:

```tcl
proc target {} { return 1 }
interp alias {} short {} target
short                            ;# tclsh: succeeds
interp alias {} short {}
```

tclsh: `short` succeeds. The analyser: W123 fires on `short` anyway, because `alias_names` only asks "is this alias's most recent action in the whole file a live definition or a deletion", ignoring where the call sits relative to the deletion.

**2. A deletion recorded inside a proc body that's never called still poisons the whole-file set**, even though the deletion never actually executes:

```tcl
proc target {} { return 1 }
interp alias {} short {} target
proc maybeDelete {} { interp alias {} short {} }
proc caller {} { short }
```

tclsh: calling `caller` (without ever calling `maybeDelete`) succeeds. The analyser: W123 fires on `short` inside `caller`, because the conditional deletion inside `maybeDelete` is treated as unconditional.

## What the fix looks like

Same shape as #973's proc/class fix: replace `alias_names: HashSet<String>` / `rename_target_names: HashSet<String>` with per-tail (or per-name) `(qualified_name, establishing_offset)` maps — mirroring `proc_defs_by_tail` / `class_defs_by_tail` — and check each candidate's liveness per call site via `fact_live_for_call` instead of `fact_live_at_file_end`, inside `w123_invocation_resolves`.

## Verification bar

- Repro tests for both gaps above (expect no W123).
- Regression tests for the cases `fact_live_at_file_end` already gets right today (alias/rename-target deleted with no re-establishment still fires W123; re-established after deletion still resolves) — don't regress these.
- All confirmed against a real tclsh (8.6 and, if available in the environment, 9.0) before/after.
- `cargo fmt --check`, `cargo clippy --workspace --all-features -- -D warnings` (no new allows), `cargo test --workspace --all-features`, `cargo xtask resolution-drift`.

## References

- Root issue: #973
- Relevant code: `rust/tcl-compiler/src/analyser/diagnostics/unresolved.rs` (`build_w123_known_names`, `fact_live_at_file_end`, `fact_live_for_call`, `w123_invocation_resolves`)

### LINKED ISSUE #1007: Arity checker (E002/E003) silently abstains when a rename/alias deletion sits inside a never-triggered proc body
## Background

While fixing #973 (W123 false negative on a deleted proc/class/alias/rename target), I found the same root-cause class of bug in the same-file arity resolver (E002/E003, `rust/tcl-compiler/src/analyser/diagnostics/validity.rs`), which uses a related but distinct mechanism: `fact_in_effect` / `fact_superseded_by_deletion`, consulted by `resolve_indirect_call_target`.

`fact_in_effect` treats a call inside a proc/method body as unconditionally seeing every top-level fact (`!cand.enforce_order || established_off < cand.call_off`) — correct for call-site order-gating. `fact_superseded_by_deletion` then reuses the *same* `fact_in_effect` check to decide whether a recorded deletion is "in effect" by the time of the call:

```rust
fn fact_superseded_by_deletion(&self, name: &str, fact_off: u32, cand: &PendingUserCallArity) -> bool {
    self.deleted_commands
        .get(name)
        .is_some_and(|&del_off| del_off > fact_off && Self::fact_in_effect(cand, del_off))
}
```

This reuse is the bug: `fact_in_effect(cand, del_off)` only asks "is the call order-gated against the *call's own* body/top-level status", not "did the deletion itself actually execute" — it has no way to know that `del_off` names a statement sitting inside a *different*, possibly-never-invoked proc body. So any `rename`/`interp alias` deletion recorded anywhere in the file — even one nested inside a proc nobody calls — permanently disqualifies the fact for every later proc-body call, causing the arity check to abstain (silently, no diagnostic) instead of checking against the still-live definition.

The W123 fix in #973 addresses the equivalent question for `unresolved.rs` via `fact_live_for_call`, which adds `self.offset_is_inside_definition_body(del_off)` — checking whether the *deletion itself* is conditional — before treating the deletion as in effect. `fact_in_effect`/`fact_superseded_by_deletion` in `validity.rs` has no equivalent check.

## Repro (confirmed against tclsh 8.6.14)

```tcl
proc target {a b} {}
proc maybeDelete {} { rename target {} }
target 1 2 3
```

tclsh: `maybeDelete` is never called, so the `rename` never executes — `target 1 2 3` fails `wrong # args: should be "target a b"`.

The analyser (`arity_codes` in `rust/tcl-compiler/src/analyser/diagnostics/tests.rs`'s test harness): reports **nothing** — no E002/E003 at all. It has silently treated `target` as permanently deleted (superseded), because `maybeDelete`'s `rename target {}` disqualifies it, and abstains rather than checking the still-live 2-arg signature.

For contrast, the call-site-order half of this already works correctly — a top-level call *before* a later, unconditional (non-body) deletion still gets checked:

```tcl
proc target {a b} {}
target 1 2 3
rename target {}
```

tclsh and the analyser agree: `E003` fires. Only the "is the deletion itself conditional" half is missing.

This affects procs, `rename` targets, and `interp alias` targets alike, since they all flow through the same `fact_superseded_by_deletion` helper via `resolve_indirect_call_target`.

## What the fix looks like

Add a check to `fact_superseded_by_deletion` (or a new helper it delegates to) mirroring `fact_live_for_call`'s `self.offset_is_inside_definition_body(del_off)` guard — a deletion recorded inside a proc/method/class body should not supersede a fact unless that body is known to run unconditionally (which it isn't, by the same "no CFG dominator model" limitation already documented elsewhere in this file for conditionally-*defined* procs).

## Verification bar

- The repro above must fire `E003` after the fix.
- The existing conditional-deletion-adjacent tests in `rust/tcl-compiler/src/analyser/diagnostics/tests.rs` (`same_file_rename_reestablished_after_deletion_checks_new_arity`, `same_file_alias_reestablished_after_deletion_checks_new_target_arity`, etc.) must not regress.
- Confirmed against a real tclsh (8.6 and, if available, 9.0) for both the fixed case and its existing-passing counterparts.
- `cargo fmt --check`, `cargo clippy --workspace --all-features -- -D warnings` (no new allows), `cargo test --workspace --all-features`.

## References

- Discovered while fixing #973 (same root-cause class, different diagnostic pass).
- Relevant code: `rust/tcl-compiler/src/analyser/diagnostics/validity.rs` (`fact_in_effect`, `fact_superseded_by_deletion`, `resolve_indirect_call_target`)

### LINKED ISSUE #1009: Constant-$cmd dispatch settlement resolves through a renamed/deleted-away proc, class, alias, or rename target with no deletion gate
## Background

Same root-cause class as #973/#1006/#1007: a "is this name a known/live command" predicate that consults `all_procs`/`all_classes`/`command_aliases`/`renamed_commands` via `.contains_key(...)` with no awareness that the name may have been renamed away or deleted (`rename NAME {}` / `interp alias {} NAME {}`, tracked in `deleted_commands`) with no later re-establishment.

Found in `rust/tcl-compiler/src/analyser/diagnostics/const_dispatch.rs`, `settle_one_site`'s `known` (lines ~111-118) and `user_defined` (lines ~119-124) closures — used to settle a `set cmd target; $cmd` (or `{*}$cmd`) style indirect dispatch site against the compiler's flow-sensitive constant-value model (M7). Neither closure checks `deleted_commands` at all (only the registry-builtin clause of `known` gates on a `renamed_away` set) — the exact shape the `known` predicate in `scope.rs`'s `finalise_invocation_resolutions` had before #973.

## Confirmed (via the analyser's own test harness, not just reasoning)

```tcl
proc target {} {}
rename target {}
set cmd target
$cmd
```

tclsh 8.6.14: `$cmd` fails `invalid command name "target"` (confirmed in earlier investigation for #973/#1006).

The analyser: the indirect invocation created for `$cmd` still resolves `resolved_qualified_name: Some("::target")`, as if `target` is still live — because `user_defined("::target")` returns `true` (no deletion check) so `settle_one_site` doesn't abstain.

## Why this does NOT cause a W123 false negative (scoped correctly)

`settle_const_dispatches` (which calls `settle_one_site`) runs from `emit_cfg_ssa_diagnostics`, which `run_diagnostic_emitters` (`rust/tcl-compiler/src/analyser/state.rs`) calls **after** `emit_unresolved_command_diagnostics` (W123) — confirmed by the code's own comment at `canonicalize_result_order` ("W123 already emitted"). So the invocations this module creates are never re-checked by W123's own (already deletion-aware, per #973/#1006) resolution logic.

**The actual impact is narrower but still real:** stale/incorrect `resolved_qualified_name` on the `SignatureCommandInvocation`s this module pushes into `result.command_invocations`, which feed hover, go-to-definition, find-references, and rename-tracking (per the module's own doc comment) for `$cmd`-style indirect dispatch sites. A `$cmd` dispatch whose value resolves to a renamed-away/deleted-with-no-reestablishment command will still show that command as its "live" target in these features, even though calling it fails at runtime.

## What the fix looks like

Extend `known`/`user_defined` in `settle_one_site` with the same offset-comparison approach `fact_live_for_call` (`unresolved.rs`) or `fact_superseded_by_deletion` (`validity.rs`) already use — checking each qualified candidate's own establishing offset against `deleted_commands`, with re-establishment and conditional-body awareness. The call-site offset available here is `site.span.start()` (the `$cmd`/`{*}$cmd` token's position).

## Verification bar

- The repro above must resolve to no target (or at least not falsely resolve as `::target`) after the fix — confirm via `command_invocations` in a unit test (see `const_cmd_head_records_a_reference_to_the_dispatched_proc_m7` in `rust/tcl-compiler/src/analyser/diagnostics/tests.rs` for the existing test shape/pattern).
- Regression guard: the same shape *with* a later re-establishment (a fresh `proc target {} {}` after the deletion) must still resolve correctly.
- Confirmed against a real tclsh (8.6 and, if available, 9.0).
- `cargo fmt --check`, `cargo clippy --workspace --all-features -- -D warnings` (no new allows), `cargo test --workspace --all-features`.

## References

- Discovered while auditing for more instances of #973's root cause after fixing #1006/#1007.
- Relevant code: `rust/tcl-compiler/src/analyser/diagnostics/const_dispatch.rs` (`settle_one_site`, `known`, `user_defined`)

### LINKED ISSUE #1010: var_command.rs has four more ungated "is this command known" checks — one actively deletes a correct W123 diagnostic
## Background

Same root-cause class as #973/#1006/#1007/(const_dispatch.rs sibling issue): an "is this name known" check against `all_procs`/`all_classes`/`command_aliases`/`renamed_commands` with no `deleted_commands` gate. `rust/tcl-compiler/src/analyser/diagnostics/var_command.rs` has **zero** references to `deleted_commands`, `fact_live_for_call`, `fact_superseded_by_deletion`, or `offset_is_inside_any_definition_body` anywhere in the file (confirmed by grep) — every "known" check in it predates the #973 fix pattern entirely. Audited every `.contains_key(...)` site in the file against real Tcl semantics; four are genuine bugs, two are not (see below).

### 1. `resolve_interpolated_w123_diagnostics` (~line 1478-1479) — highest priority: actively deletes a correct diagnostic

This closure retroactively **removes an already-emitted W123** for an interpolated command head (`do${suffix}`) once SCCP folds `$suffix` to a constant, checking the folded name against `all_procs.contains_key` with no deletion gate.

Repro (same shape confirmed working for #973/#1006):
```tcl
proc doThing {} {}
rename doThing {}
set suffix Thing
do${suffix}
```
tclsh 8.6.14: fails `invalid command name "doThing"`. The analyser: W123 is emitted initially for the unresolved interpolated head, then this closure sees the SCCP-folded name `doThing`, finds it in `all_procs` (deleted-but-not-pruned), and **deletes the W123 diagnostic outright** — a duplicate, independently-implemented "is this proc known" that never got the #973 fix, and the only one of these four that regresses an already-correct diagnostic rather than just failing to add one.

### 2. `is_known_command` (~line 578), used by `w307_site_suppressed`

Gates suppression of W307 ("non-literal command name — cannot statically analyze") once SCCP resolves a `$var` dispatch head to a constant. No deletion gate, and no call-site offset is even threaded into the closure today, so a `fact_live_for_call`-style fix needs that plumbed through `w307_site_suppressed` first. Not redundantly covered elsewhere — this is `var_command.rs`'s own territory (dynamic `$var` dispatch), so a dead-but-still-`all_procs`-resident name here genuinely suppresses a warning real Tcl's runtime would reject.

### 3. `emit_dispatch_table_command_references`'s `known` (~line 1266-1269)

Synthesizes a `SignatureCommandInvocation` reference for a dispatch-table literal (`array set`/`dict` proc-name string) when its value is a known proc/class — feeding rename-tracking, references, and (via `command_invocations`) call-graph/dead-code consumers. Confirmed by code-flow (not just reasoning): this runs from `emit_cfg_ssa_diagnostics` → `emit_var_command_diagnostics`, which — like the sibling `const_dispatch.rs` issue — executes *after* `finalise_invocation_resolutions` (`scope.rs`, the #973-fixed, deletion-aware settling pass) already ran, so entries this closure pushes are never re-settled and permanently keep whatever the ungated `known` decided. Lower urgency than #1/#2 (feeds navigation/rename-tracking, not a diagnostic code directly), but same fix shape.

### 4. `harvest_constructor_object_types` (~line 144-147) / `emit_cmd_command_diagnostics` (~line 1070-1071) — same check, two call shapes

Checks whether `head`/`class_qn` is a known `all_classes` entry to attribute `set x [Cls new]` (or `[Cls new] method`) as producing an `Object(Cls)`, driving W308 method-name validation. Lowest priority: if `Cls` was renamed away with no re-establishment, real Tcl fails the constructor call itself with "invalid command name" — which the already-fixed, deletion-aware class path in `unresolved.rs`'s W123 very likely already catches as the primary diagnostic. This gap mostly produces a redundant/misleading secondary W308-family diagnostic rather than the sole miss, and the trigger (renaming a TclOO class command) is rare. Worth fixing for consistency if the area is being touched anyway, not urgent standalone.

## Confirmed NOT bugs (different questions, no gate needed)

- **`compute_factory_object_ranges`'s `is_user_proc`** (~line 706-709, used at ~723) — this only *routes* factory-object classification (name-shape heuristic vs. the separate proc-return-value fixpoint); it doesn't itself answer the terminal "is this class currently live" question, so gating it changes nothing about correctness.
- **`has_external_super_or_mixin`** (~line 1368-1373) — asks "do I have structural knowledge of this class's method table" (an inheritance-edge/MRO fact established once, at `superclass`/`mixin` declaration time), not "is this class's command currently callable." TclOO's real semantics: renaming or deleting a class command after a subclass link is established does not invalidate that subclass's inherited method table (the well-known "rename the class away after subclassing, existing subclasses/instances keep working" idiom) — so the ungated check matches real behavior here.

## Verification bar

- Repro tests for each of the four bugs above (using the same tclsh-confirmed pattern already established for #973/#1006/#1007), expecting the corrected behavior after fixing.
- Regression guards for the re-established-after-deletion case for each.
- Confirmed against a real tclsh (8.6 and, if available, 9.0).
- `cargo fmt --check`, `cargo clippy --workspace --all-features -- -D warnings` (no new allows), `cargo test --workspace --all-features`.

## References

- Discovered while auditing for more instances of #973's root cause after fixing #1006/#1007.
- Relevant code: `rust/tcl-compiler/src/analyser/diagnostics/var_command.rs`

### PR DESCRIPTION
**Pull request by bitwisecook**
## Summary

- Fixes #973: `finalise_invocation_resolutions`'s `known` predicate (`scope.rs`) and the real W123-firing logic (`unresolved.rs`) never gated proc/class/alias/rename-target resolution on `rename NAME {}` / `interp alias {} NAME {}` deletion — a deleted-with-no-re-establishment name still resolved as callable.
- Fixes #1006: extends the same fix to `unresolved.rs`'s alias/rename-target checks, making them call-site- and conditional-body-aware (matching the sophistication already used for registry builtins) instead of file-end-only.
- Fixes #1007: the arity checker's (`validity.rs`) `fact_superseded_by_deletion` reused a call-site-order helper to also decide whether a *deletion* was in effect, with no way to know the deletion sat inside a never-triggered proc body — so any `rename`/`interp alias` deletion anywhere in the file permanently (and wrongly) silenced E002/E003 for later calls.
- Fixes #1009: constant-`$cmd` dispatch settlement (`const_dispatch.rs`) resolved through a renamed/deleted command with no gate at all, poisoning hover/go-to-definition/references/rename-tracking for indirect dispatch sites (confirmed this never caused a W123 false negative, since that pass runs after W123 already fires).
- Fixes #1010: four more ungated checks in `var_command.rs` — interpolated-W123 resolution (was actively **deleting** an already-correct W123), W307 dynamic-dispatch suppression, dispatch-table reference synthesis, and TclOO constructor object-typing (was producing a misleading secondary W308 instead of the real W123 + a conservative W307 fallback).
- Fixes two false positives found in Codex review of this PR (both confirmed against tclsh 8.6.14):
  - `finalise_invocation_resolutions` picked between a local and a global candidate using a file-end-only deletion check, so a namespaced local call textually before a later unconditional deletion wrongly lost to the global candidate.
  - `fact_live_for_call` treated any call inside a proc/class body as automatically after every top-level deletion, drawing a spurious W123 even when the enclosing definition's own top-level invocation demonstrably ran before that deletion.

  Both fixed via a shared escape hatch: track the earliest top-level (non-body) call-site offset per resolved qualified name, and when a call is nested inside a body, check whether its innermost enclosing definition has a provable top-level invocation before the deletion.
- Also fixes a VS Code extension test that started failing in CI: it compared reference results without filtering by document URI, so a sibling fixture left open by an earlier test in the same suite (same proc name, minus the deletion) was leaking legitimate cross-document references into the assertion.

All fixes reuse a single mechanism (`fact_live_for_call` / `fact_live_at_file_end` / `group_defs_by_tail`, widened to `pub(super)` so sibling passes share one implementation) rather than re-deriving the same "is this name's most recent fact a live definition or a deletion" question five different ways. Every case — TP (deleted, no re-establishment), FP guard (re-established after deletion), TN (never deleted), and the call-site/conditional-body edge cases — was verified against a real tclsh 8.6.14 before being encoded as a test, not assumed.

Two residuals, deliberately **not** folded into this PR:
- `set x [Cls new]` (the variable-assignment constructor shape) still has an independent, ungated source in `type_infer.rs`'s SSA type-lattice construction — a foundational, flow-insensitive module with no call-site offset available today. Tracked separately as #1013.
- The body-call escape hatch added in response to Codex review only checks one level of enclosing-definition indirection (`proc outer {} { inner }` calling `proc inner {} { helper }` where only `outer` has a top-level call site) — a fully correct fix needs call-graph reachability with cycle handling, a bigger and riskier change than the two confirmed findings warranted rushing in. Tracked separately as #1015.

## Validation

- `cargo test --workspace --all-features` — full workspace, including the native `lsp-e2e` suite (`rust/tcl-lsp-server/tests/e2e/`).
- `cargo fmt --all --check`
- `cargo clippy --workspace --all-features -- -D warnings` (no new allows; several functions needed line-count-driven helper extraction to stay compliant)
- `cargo xtask resolution-drift` (the namespace-blind `.name ==` scan gate — relevant since this PR adds new `all_procs`/`all_classes` scans)
- VS Code extension tests added/fixed (`editors/vscode/src/test/issue945.test.ts`, `precisionReview.test.ts`) — run for real against the actual Electron-hosted test host (not just type-checked), confirming both new tests and the previously-failing CI test now pass.
- Found and fixed two genuine pre-existing regression risks along the way: a VS Code test (`precisionReview.test.ts`) and a native e2e test (`precision_review.rs`) both asserted "zero W123 anywhere in this fixture," an assumption that only held because of #973's bug — both updated to assert the correct, scoped behavior; and a VS Code reference-provider test that leaked cross-document results from a sibling fixture left open by an earlier test in the same suite.

## Compiler/KCS checklist

- [x] Did this change alter a compiler fact contract? Yes — W123/E002/E003 now fire in more cases (previously silent false negatives), and resolve correctly in more cases (previously silent false positives from the Codex review round).
- [x] Updated `docs/kcs/codes/kcs-diagnostic-w123-unresolved-command.md` with the newly-covered shape (a renamed/deleted-away command with no re-establishment).
- [x] Updated `README.md`'s Diagnostics section with the renamed/deleted-command W123 behaviour, per AGENTS.md's documentation requirement (flagged in Codex review).

## Follow-ups tracked separately

- #1013 — the deeper `type_infer.rs` gap noted above.
- #1015 — the one-level-only body-call escape hatch noted above.

### CODE REVIEW

File: rust/tcl-compiler/src/analyser/scope.rs [hunk 1]
```diff
@@ -412,31 +429,54 @@ impl Analyser {
...
+        // re-established under the same name (a fresh `proc`, `rename`, or
+        // `interp alias`) is live again, so only a deletion that postdates
+        // this specific fact's own establishing offset disqualifies it.
+        // `deleted_commands` stores only the last-seen deletion offset per
+        // name, which — since the walk visits statements in source order —
+        // is always the most recent one.
+        let live = |qualified: &str, fact_off: u32| {
+            deleted_commands
+                .get(qualified)
+                .is_none_or(|&del_off| del_off <= fact_off)
```
[the remark this answers is not in the document]
Comment (bitwisecook) [reply rust/tcl-compiler/src/analyser/scope.rs]: Confirmed against tclsh 8.6.14 and fixed in fde3225: `finalise_invocation_resolutions`'s `known` predicate is now call-site aware (threads `call_off` through), with a new escape hatch (`top_level_call_offsets`) for the case where the call itself sits inside a body whose enclosing definition has a provable top-level invocation before the deletion. See `KnownPredicateCtx::live_for_call` in `scope.rs`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: rust/tcl-compiler/src/analyser/diagnostics/unresolved.rs [hunk 2]
```diff
@@ -150,6 +187,100 @@ impl Analyser {
...
+        let Some(&del_off) = self.deleted_commands.get(qualified) else {
+            return true;
+        };
+        if del_off <= fact_off {
+            return true;
+        }
+        if self.offset_is_inside_definition_body(del_off) {
+            return true;
+        }
+        !self.offset_is_inside_definition_body(call_off) && call_off <= del_off
```
[the remark this answers is not in the document]
Comment (bitwisecook) [reply rust/tcl-compiler/src/analyser/diagnostics/unresolved.rs]: Confirmed against tclsh 8.6.14 and fixed in fde3225: `fact_live_for_call` now falls back to checking whether the call's innermost enclosing definition has its own top-level invocation before the deletion, rather than treating every body call as unconditionally after every top-level deletion. Your exact repro (`proc helper {} {}; proc caller {} { helper }; caller; rename helper {}`) is now a permanent regression test (`w123_fp_issue_1009_codex_review_body_call_before_later_deletion_resolves`).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: docs/kcs/codes/kcs-diagnostic-w123-unresolved-command.md [hunk 3]
```diff
@@ -31,6 +31,21 @@ unknownCmd $arg
 
 The analyser reports **`W123`** on `unknownCmd`.
 
+A proc, class, `rename` target, or `interp alias` that was renamed or
+deleted away, with no later re-establishment under the same name, is also
+unresolved — calling it fails `invalid command name` at runtime just like
```
[the remark this answers is not in the document]
Comment (bitwisecook) [reply docs/kcs/codes/kcs-diagnostic-w123-unresolved-command.md:36]: Added a README.md note (Diagnostics section) on the renamed/deleted-command W123 behaviour in fde3225.

---
_Generated by [Claude Code](https://claude.ai/code)_
